src/librustc/infer/error_reporting.rs

```diff
@@ -565,19 +565,29 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
                 // look for expected with found id
                 self.tcx.populate_inherent_implementations_for_type_if_necessary(found);
                 if let Some(impl_infos) = self.tcx.inherent_impls.borrow().get(&found) {
-                    let mut methods = Vec::new();
```

Comment:
Can you write like this:
```rust
let methods = impl_infos.iter()
                        .flat_map(|impl_| {
                          self.tcx
                              .impl_or_trait_items(*impl_)
                              .iter()
                              .map(|&did| (None, did, self.tcx.impl_or_trait_item(did)))
                              .filter(|&(_, _, ref x)| {
                                self.matches_return_type(x, &expected_ty)
                              })
                        });
```
I didn't test it, so it may be impossible.